assets/ts/scrollspy.ts

```diff
@@ -0,0 +1,125 @@
+    if (scrollTop < 0) {
+        scrollTop = 0;
+    }
+    scrollableNavigation.scrollTo({ top: scrollTop, behavior: "smooth" });
+}
+
+function findLinkForSectionId(sectionId: string, navigation: NodeListOf<Element>): HTMLElement | undefined {
+    for (let i = 0; i < navigation.length; i++) {
+        let link = navigation[i].querySelector("a");
+        if (link.getAttribute("href") === "#" + sectionId) {
+            return navigation[i] as HTMLElement;
+        }
+    }
+
+    return undefined;
+}
+
+function computeOffsets(headers: NodeListOf<Element>) {
+    let sectionsOffsets = [];
+    headers.forEach((header: HTMLElement) => { sectionsOffsets.push({ id: header.id, offset: header.offsetTop }) });
+    sectionsOffsets.sort((a, b) => a.offset - b.offset);
+    return sectionsOffsets;
+}
+
+function setupScrollspy() {
+    let headers = document.querySelectorAll(headersQuery);
+    if (!headers) {
+        console.warn("No header matched query", headers);
+        return;
+    }
+
+    let scrollableNavigation = document.querySelector(tocQuery) as HTMLElement | undefined;
+    if (!scrollableNavigation) {
+        console.warn("No toc matched query", tocQuery);
+        return;
+    }
+
+    let navigation = document.querySelectorAll(navigationQuery);
+    if (!navigation) {
+        console.warn("No navigation matched query", navigationQuery);
+        return;
+    }
+
+    let sectionsOffsets = computeOffsets(headers);
+
+    // We need to avoid scrolling when the user is actively interacting with the ToC. Otherwise, if the user clicks on a link in the ToC,
+    // we would scroll their view, which is not optimal usability-wise.
+    let tocHovered: boolean = false;
+    scrollableNavigation.addEventListener("mouseenter", debounced(() => tocHovered = true));
+    scrollableNavigation.addEventListener("mouseleave", debounced(() => tocHovered = false));
+
+    let activeSectionLink: Element;
+
+    function scrollHandler() {
+        let scrollPosition = document.documentElement.scrollTop || document.body.scrollTop;
+
+        let newActiveSection: HTMLElement | undefined;
+
+        // Find the section that is currently active.
+        // It is possible for no section to be active, so newActiveSection may be undefined.
+        sectionsOffsets.forEach((section) => {
+            if (scrollPosition >= section.offset - 20) {
+                newActiveSection = document.getElementById(section.id);
+            }
+        });
+
+        // Find the link for the active section. Once again, there are a few edge cases:
+        // - No active section = no link => undefined
+        // - No active section but the link does not exist in toc (e.g. because it is outside of the applicable ToC levels) => undefined
+        let newActiveSectionLink: HTMLElement | undefined 
+        if (newActiveSection) {
+            newActiveSectionLink = findLinkForSectionId(newActiveSection.id, navigation);
```

Comment:
Instead of iterating HTML each time, since the TOC element does not change dynamically, it might be more efficient to build a map (id -> link ref) outside `scrollHandler`:

```js
const sectionLinkRef: { [key: string]: HTMLElement } = {};
navigation.forEach((navigationElement: HTMLElement) => {
    const link = navigationElement.querySelector("a");
    const id = link.getAttribute("href").replace('#', '');
    sectionLinkRef[id] = navigationElement;
});
```

And accessing it by:
```typescript
if (newActiveSection) {
    /// Undefined if id does not exist
    newActiveSectionLink = sectionLinkRef[newActiveSection.id];
}
```

How does that sound to you?

Reply:
Good idea, I didn't think of that 👍🏻

Reply:
Done, let me know what you think. I didn't go with the exact same code you provided, but it does the same thing.